src/network/bufferedconnection.cc

```diff
@@ -0,0 +1,374 @@
+	for (size_t i = peek_pointer_; i < conn_->receive_buffer_.size(); ++i) {
+		if (conn_->receive_buffer_[i] == '\0') {
+			peek_pointer_ = i + 1;
+			return true;
+		}
+	}
+	return false;
+}
+
+bool BufferedConnection::Peeker::cmd(RelayCommand* out) {
+
+	std::lock_guard<std::mutex> lock(conn_->mutex_receive_);
+
+	assert(conn_->receive_buffer_.size() >= peek_pointer_);
+
+	if (conn_->receive_buffer_.size() > peek_pointer_) {
+		if (out != nullptr) {
+			*out = static_cast<RelayCommand>(conn_->receive_buffer_[peek_pointer_]);
+		}
+		peek_pointer_++;
+		return true;
+	}
+	return false;
+}
+
+bool BufferedConnection::Peeker::uint8_t(::uint8_t* out) {
+
+	std::lock_guard<std::mutex> lock(conn_->mutex_receive_);
+
+	assert(conn_->receive_buffer_.size() >= peek_pointer_);
+
+	// If there is any byte available, we can read an uint8
+	if (conn_->receive_buffer_.size() > peek_pointer_) {
+		if (out != nullptr) {
+			*out = static_cast<::uint8_t>(conn_->receive_buffer_[peek_pointer_]);
+		}
+		peek_pointer_++;
+		return true;
+	}
+	return false;
+}
+
+bool BufferedConnection::Peeker::recvpacket() {
+
+	std::lock_guard<std::mutex> lock(conn_->mutex_receive_);
+
+	assert(conn_->receive_buffer_.size() >= peek_pointer_);
+
+	if (conn_->receive_buffer_.size() < peek_pointer_ + 2) {
+		// Not even enough space for the size of the recvpacket
+		return false;
+	}
+
+	// RecvPackets have their size coded in their first two bytes.
+	// See SendPacket in network.cc
+	const uint16_t size = conn_->receive_buffer_[peek_pointer_ + 0] << 8 | conn_->receive_buffer_[peek_pointer_ + 1];
+	assert(size >= 2);
+
+	if (conn_->receive_buffer_.size() >= peek_pointer_ + size) {
+		peek_pointer_ += size;
+		return true;
+	}
+	return false;
+}
+
+std::unique_ptr<BufferedConnection> BufferedConnection::connect(const NetAddress& host) {
+	assert(host.is_valid());
+	std::unique_ptr<BufferedConnection> ptr(new BufferedConnection(host));
+	if (!ptr->is_connected()) {
+		ptr.reset();
+	}
+
+	return ptr;
+}
+
+std::unique_ptr<BufferedConnection> BufferedConnection::accept(boost::asio::ip::tcp::acceptor& acceptor) {
+	assert(acceptor.is_open());
+	std::unique_ptr<BufferedConnection> ptr(new BufferedConnection(acceptor));
+	if (!ptr->is_connected()) {
+		ptr.reset();
+	}
+	return ptr;
+}
+
+BufferedConnection::~BufferedConnection() {
+	close();
+	assert(!asio_thread_.joinable());
+}
+
+bool BufferedConnection::is_connected() const {
+	return socket_.is_open();
+}
+
+void BufferedConnection::close() {
+	if (!is_connected()) {
+		return;
+	}
+	boost::system::error_code ec;
+	boost::asio::ip::tcp::endpoint remote = socket_.remote_endpoint(ec);
+	if (!ec) {
+		log("[BufferedConnection] Closing network socket connected to %s:%i.\n",
+		    remote.address().to_string().c_str(), remote.port());
+	} else {
+		log("[BufferedConnection] Closing network socket.\n");
+	}
+	// Stop the thread
+	io_service_.stop();
+	while (!io_service_.stopped()); // Not sure if that is required
```

Comment:
Not sure is this could result in some endless loop.

Reply:
Good point, I don't know either. I transformed it to a for-loop, just in case.